pkg/proc/bininfo.go

```diff
@@ -1495,6 +1496,84 @@ func (bi *BinaryInfo) parseDebugFrameMacho(image *Image, exe *macho.File, debugI
 	bi.parseDebugFrameGeneral(image, debugFrameBytes, "__debug_frame", debugFrameErr, ehFrameBytes, ehFrameAddr, "__eh_frame", frame.DwarfEndian(debugInfoBytes))
 }
 
+// macOSDebugFrameBugWorkaround applies a workaround for:
+//  https://github.com/golang/go/issues/25841
+// It finds the Go function with the lowest entry point and the first
+// debug_frame FDE, calculates the difference between the start of the
+// function and the start of the FDE and sums it to all debug_frame FDEs.
+// A number of additional checks are performed to make sure we don't ruin
+// executables unaffected by this bug.
+func (bi *BinaryInfo) macOSDebugFrameBugWorkaround() {
+	//TODO: log extensively because of bugs in the field
+	if bi.GOOS != "darwin" || bi.Arch.Name != "arm64" {
+		return
+	}
+	if len(bi.Images) > 1 {
+		// Only do this for the first executable, but it might work for plugins as
+		// well if we had a way to distinguish where entries in bi.frameEntries
+		// come from
+		return
+	}
+	exe, ok := bi.Images[0].closer.(*macho.File)
+	if !ok {
+		return
+	}
+	if exe.Flags&macho.FlagPIE == 0 {
+		bi.logger.Infof("debug_frame workaround not needed: not a PIE (%#x)", exe.Flags)
+		return
+	}
+
+	// Find first Go function (first = lowest entry point)
+	var fn *Function
+	for i := range bi.Functions {
+		if bi.Functions[i].cu.isgo && bi.Functions[i].Entry > 0 {
```

Comment:
maybe need additional `bi.Functions[i].cu != nil && ..` check?

when I debug with https://github.com/cockroachdb/cockroach 's main method, it will NPE on this line

change to `if bi.Functions[i].cu != nil && bi.Functions[i].cu.isgo && bi.Functions[i].Entry > 0 {`, it works for me 😄

Reply:
@lysu that shouldn't happen, can you upload an executable that reproduces this issue somewhere?

Reply:
@aarzilli sorry for late reponse, I had sent a email linked with google driver zip file from (suli#p...com), thanks

Reply:
and run with those commands to reproduce

```
./dlv --listen=0.0.0.0:65356 --headless=true --api-version=2 --check-go-version=false --only-same-user=false exec binary_for_segv
API server listening at: [::]:65356
debugserver-@(#)PROGRAM:LLDB  PROJECT:lldb-1200.0.44
 for arm64.
Got a connection, launched process binary_for_segv (pid = 38644).
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x40 pc=0x1002fce74]

goroutine 1 [running]:
github.com/go-delve/delve/pkg/proc.(*BinaryInfo).macOSDebugFrameBugWorkaround(0x1400015a140)
	/Users/robi/Code/go/src/github.com/go-delve/delve/pkg/proc/bininfo.go:1529 +0xf4
github.com/go-delve/delve/pkg/proc.(*BinaryInfo).AddImage(0x1400015a140, 0x16fd8b841, 0xf, 0x10023c000, 0x0, 0x0)
	/Users/robi/Code/go/src/github.com/go-delve/delve/pkg/proc/bininfo.go:707 +0x290
github.com/go-delve/delve/pkg/proc.(*BinaryInfo).LoadBinaryInfo(0x1400015a140, 0x16fd8b841, 0xf, 0x10023c000, 0x140001ca560, 0x1, 0x1, 0x14000218300, 0x1004e4f7c)
	/Users/robi/Code/go/src/github.com/go-delve/delve/pkg/proc/bininfo.go:452 +0x8c
github.com/go-delve/delve/pkg/proc.NewTarget(0x1006ba598, 0x1400015a280, 0x1006b9618, 0x140001b60d0, 0x16fd8b841, 0xf, 0x140001ca560, 0x1, 0x1, 0x10101, ...)
	/Users/robi/Code/go/src/github.com/go-delve/delve/pkg/proc/target.go:151 +0x98
github.com/go-delve/delve/pkg/proc/gdbserial.(*gdbProcess).initialize(0x1400015a280, 0x16fd8b841, 0xf, 0x140001ca560, 0x1, 0x1, 0x14000218201, 0x3, 0x3, 0x0)
	/Users/robi/Code/go/src/github.com/go-delve/delve/pkg/proc/gdbserial/gdbserver.go:651 +0x188
github.com/go-delve/delve/pkg/proc/gdbserial.(*gdbProcess).Connect(0x1400015a280, 0x1006b8948, 0x14000130018, 0x16fd8b841, 0xf, 0x0, 0x140001ca560, 0x1, 0x1, 0x1, ...)
	/Users/robi/Code/go/src/github.com/go-delve/delve/pkg/proc/gdbserial/gdbserver.go:310 +0x23c
github.com/go-delve/delve/pkg/proc/gdbserial.(*gdbProcess).Listen(0x1400015a280, 0x1006b1e30, 0x1400000e0f0, 0x16fd8b841, 0xf, 0x0, 0x140001ca560, 0x1, 0x1, 0x1, ...)
	/Users/robi/Code/go/src/github.com/go-delve/delve/pkg/proc/gdbserial/gdbserver.go:253 +0x14c
github.com/go-delve/delve/pkg/proc/gdbserial.LLDBLaunch(0x14000132d80, 0x1, 0x6, 0x1004e37c2, 0x1, 0x1000e3801, 0x140001ca560, 0x1, 0x1, 0x0, ...)
	/Users/robi/Code/go/src/github.com/go-delve/delve/pkg/proc/gdbserial/gdbserver.go:506 +0x9e0
github.com/go-delve/delve/service/debugger.(*Debugger).Launch(0x140001ee8c0, 0x14000132d80, 0x1, 0x6, 0x1004e37c2, 0x1, 0x1, 0x100081d08, 0x14000132f00)
	/Users/robi/Code/go/src/github.com/go-delve/delve/service/debugger/debugger.go:291 +0x4a0
github.com/go-delve/delve/service/debugger.New(0x140001ba1c0, 0x14000132d80, 0x1, 0x6, 0x0, 0x0, 0x0)
	/Users/robi/Code/go/src/github.com/go-delve/delve/service/debugger/debugger.go:184 +0x498
github.com/go-delve/delve/service/rpccommon.(*ServerImpl).Run(0x14000132f00, 0x14000132f00, 0x1400010a3c0)
	/Users/robi/Code/go/src/github.com/go-delve/delve/service/rpccommon/server.go:115 +0xb8
github.com/go-delve/delve/cmd/dlv/cmds.execute(0x0, 0x14000132d80, 0x1, 0x6, 0x140001c31e0, 0x0, 0x0, 0x0, 0x14000132d80, 0x1, ...)
	/Users/robi/Code/go/src/github.com/go-delve/delve/cmd/dlv/cmds/commands.go:860 +0x688
github.com/go-delve/delve/cmd/dlv/cmds.New.func4(0x140001f06c0, 0x14000132d80, 0x1, 0x6)
	/Users/robi/Code/go/src/github.com/go-delve/delve/cmd/dlv/cmds/commands.go:229 +0x78
github.com/spf13/cobra.(*Command).execute(0x140001f06c0, 0x140001ee780, 0x6, 0xa, 0x140001f06c0, 0x140001ee780)
	/Users/robi/Code/go/pkg/mod/github.com/spf13/cobra@v0.0.0-20170417170307-b6cb39589372/command.go:647 +0x18c
github.com/spf13/cobra.(*Command).ExecuteC(0x1400014fb00, 0x1400014fb00, 0x1004e6134, 0x6)
	/Users/robi/Code/go/pkg/mod/github.com/spf13/cobra@v0.0.0-20170417170307-b6cb39589372/command.go:733 +0x208
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/robi/Code/go/pkg/mod/github.com/spf13/cobra@v0.0.0-20170417170307-b6cb39589372/command.go:692
main.main()
	/Users/robi/Code/go/src/github.com/go-delve/delve/cmd/dlv/main.go:24 +0xc0
Exiting.
```